core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -29,10 +29,15 @@ final case class Uri(
   fragment: Option[Fragment] = None)
   extends QueryOps with Renderable
 {
+  import Uri._
+
   def withPath(path: Path): Uri = copy(path = path)
 
-  def /(newFragment: Path): Uri = {
-    val newPath = if (path.isEmpty || path.last != '/') path + "/" + newFragment else path + newFragment
+  def /(newSegment: Path)(implicit charset: Charset = Charset.`UTF-8`): Uri = {
```

Comment:
I can't think of it ever making sense to change charsets while constructing a URI.  Making it a property of the URI is possible, but there are a lot of URIs, and 99.9999% of them are UTF-8 and the other 0.0001% are bad and should feel bad.

A richer data type would allow deferring the encoding to a time that one charset is passed in.  The smarter data type is proposed at #559